core/src/xtdb/checkpoint.clj

```diff
@@ -142,24 +152,29 @@
         cp)))
 
   (download-checkpoint [_ {::keys [cp-path]} dir]
-    (let [to-path (.toPath ^File dir)]
+    (let [to-path (.toPath ^File dir)
+          success (atom 0)]
       (when-not (or (not (Files/exists to-path (make-array LinkOption 0)))
                     (empty? (Files/list to-path)))
         (throw (IllegalArgumentException. "non-empty checkpoint restore dir: " to-path)))
 
       (try
         (sync-path cp-path to-path)
+        (swap! success inc)
```

Comment:
I don't think you need any state in here - IIUC you're essentially doing this?

```clojure
(try
  (upload-checkpoint!)
  (catch Throwable t
    (cleanup-checkpoint!)
    (throw (ex-info "incomplete checkpoint" {...} t))))
```

Reply:
subjective style note (feel free to keep it as it is, it works perfectly fine) but I'd probably have the cleanup included in the contract of each `upload-checkpoint` rather than introduce a separate protocol function?

It's quite a common pattern for us to have various functions shaped like the following, so that most functions have an implicit contract that, if they go wrong, the caller can trust that they don't leave anything behind:

```clojure
(try
  ;; do the thing
  (catch Throwable t
    ;; clean up
    (throw t)))
```

Reply:
maybe different on the download case because the caller 'owns' the temporary directory, and is instructing the cp-store to use it - i.e. the cleanup's the same for all of the implementations

Reply:
When I added  `cleanup-checkpoint` to the protocol, I had in mind to indeed enrich the protocol, e.g., having also a `valid-checkpoint?` method (and perhaps other), to allow us to implement some checkpoint management procedure which would operate in an abstract manner (i.e., regardless the kind of store it is using).